### PR TITLE
[Transforms] Resolve FIXME: No Lowering of FNeg to FMul unless it is safe

### PR DESCRIPTION
**Pull request by AZero13**
Or association flags are specified.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Could you please provide the proof? I cannot prove the correctness with alive2.

**Comment by dtcxzyw** [comment]
https://alive2.llvm.org/ce/z/Ki3Rf2

`nnan` is all you need :)

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/test/Transforms/Reassociate/fast-basictest.ll [hunk 1]
```diff
@@ -578,3 +578,61 @@ define float @test18(float %A, float %B) {
...
+; CHECK-LABEL: @scalar(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[NEG:%.*]] = fmul fast float [[A:%.*]], 2.000000e+00
+; CHECK-NEXT:    [[MUL:%.*]] = fmul fast float [[NEG]], [[A]]
+; CHECK-NEXT:    ret float [[MUL]]
+;
+entry:
+  %b = fmul fast float %a, -2.0
+  %neg = fneg fast float %a
+  %mul = fmul fast float %neg, %b
```
Comment (arsenm) [inline llvm/test/Transforms/Reassociate/fast-basictest.ll:594]: Don't need the full set of fast flags. Also don't need it on every instruction. Should target testing missing specific flags contexts

File: llvm/test/Transforms/Reassociate/fast-basictest.ll [hunk 2]
```diff
@@ -578,3 +578,61 @@ define float @test18(float %A, float %B) {
...
+; CHECK-NEXT:    [[NEG:%.*]] = fmul fast <4 x float> [[A:%.*]], <float -1.000000e+00, float -2.000000e+00, float 3.000000e+00, float -4.000000e+00>
+; CHECK-NEXT:    [[MUL:%.*]] = fmul fast <4 x float> [[NEG]], [[A]]
+; CHECK-NEXT:    ret <4 x float> [[MUL]]
+;
+entry:
+  %b = fmul fast nnan <4 x float> %a, <float 1.0, float 2.0, float -3.0, float 4.0>
+  %neg = fneg fast nnan <4 x float> %a
+  %mul = fmul fast <4 x float> %neg, %b
+  ret <4 x float> %mul
+}
```
Comment (arsenm) [inline llvm/test/Transforms/Reassociate/fast-basictest.ll:638]: Also add test where it's know not-nan from the source

**Comment by arsenm** [comment]
> https://alive2.llvm.org/ce/z/Ki3Rf2
> 
> `nnan` is all you need :)

In this context, it's not simply doing this transform in isolation. It's being consumed by another canonicalizing operation. The problem is this is introducing a canonicalization, which doesn't matter if it's only used by operations that canonicalize the inputs. nnan would not be sufficient to introduce this, since operations may also do other bit pattern changing side effects (like denormal flush). Since this is consumed by a canonicalizing operation, it's either correct to always do (in the has one use case) (or it's wrong to ever do).

**Review by arsenm — CHANGES_REQUESTED** [review]

**Comment by arsenm** [comment]
> which doesn't matter if it's only used by operations that canonicalize the inputs. 

I guess we technically don't guarantee canonicalization on FP operations, but I think we can assert that it may happen

**Pull request closed without merging**